source/common/http/filter_manager.cc

```diff
@@ -1064,6 +1064,19 @@ void FilterManager::encodeHeaders(ActiveStreamEncoderFilter* filter, ResponseHea
     }
   }
 
+  const auto status = HeaderUtility::checkRequiredResponseHeaders(headers);
+  if (!status.ok()) {
+    // This branch can happen when a misbehaving filter chain removed critical headers or set
+    // mulformed header values.
+    const auto request_headers = filter_manager_callbacks_.requestHeaders();
+    sendLocalReply(request_headers.has_value() &&
+                       Grpc::Common::isGrpcRequestHeaders(request_headers.ref()),
+                   Http::Code::InternalServerError, status.message(), nullptr, absl::nullopt,
+                   absl::StrCat(StreamInfo::ResponseCodeDetails::get().FilterRemovedRequiredHeaders,
+                                "{", status.message(), "}"));
```

Comment:
followed the request path here: https://github.com/envoyproxy/envoy/blob/main/source/common/router/upstream_request.cc#L444-L453